### PR TITLE
Added permission for token creation

### PR DESCRIPTION
**Pull request by OliviaYtterbrink**
Added check after we know that a token needs to be created.
Default role needed to create tokens is `Architect`, can change `dbms.security.allow_publisher_create_token` to allow `Publisher` as well.

changelog: [security] Added new permission level for handling creation of tokens

* Default role given this permission is `Architect` and `Admin`
* setting `dbms.security.allow_publisher_create_token` to true will give this permission to `Publisher` as well

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java [hunk 1]
```diff
@@ -173,6 +197,46 @@ public boolean allowsWrites()
...
+                        return true;
+                    }
+
+                    @Override
+                    public boolean allowsTokenCreates()
+                    {
+                        return true;
+                    }
+
+                    @Override
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java]: A note on style: I prefer to pass values through the constructor to fields of the enums, over the proliferation of overridden abstract methods. It makes the class shorter, since the getters only have to be implemented once.

File: community/kernel/src/main/java/org/neo4j/kernel/api/proc/Mode.java [hunk 2]
```diff
@@ -29,6 +29,8 @@
     READ_ONLY,
     /** This procedure may perform both read and write operations against the graph */
     READ_WRITE,
+    /** This procedure may perform both read and write operations against the graph and create new tokens */
+    TOKEN,
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/api/proc/Mode.java]: I find it strange that the ability to create tokens goes beyond write. At first glance, I thought this was about allowing read-only transactions to also create tokens, but apparently not? Can you explain why TOKEN is a super-set of READ_WRITE?

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/TokenProcedures.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            throws IllegalTokenNameException    {
+        tx.acquireStatement().tokenWriteOperations().relationshipTypeGetOrCreateForName( newRelationshipType );
+    }
+
+    @Description( "Create a Property" )
+    @Procedure( name = "db.createProperty", mode = TOKEN )
+    public void createProperty( @Name( "newProperty" ) String newProperty ) throws IllegalTokenNameException
+    {
+        tx.acquireStatement().tokenWriteOperations().propertyKeyGetOrCreateForName( newProperty );
+    }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/TokenProcedures.java:58]: We need this so a DBA can prepare the schema items in advance (for people with READ and READ_WRITE modes, presumably) without adding any data to the database?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java [hunk 4]
```diff
@@ -135,6 +135,7 @@ final KeyReadOperations tokenRead()
 
     final KeyWriteOperations tokenWrite()
     {
+        statement.assertAllows( AccessMode::allowsTokenCreates, "Token create" );
         return operations.keyWriteOperations();
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:139]: This `"Token create"` shouldn't really be hard-coded like this, should it?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/Procedures.java [hunk 5]
```diff
@@ -217,4 +218,14 @@ public void start() throws Throwable
         // And register built-in procedures
         builtin.accept( this );
     }
+
+    public void setAllowWriteTokenCreate( boolean allowWriteTokenCreate )
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/Procedures.java]: Could this (field, parameter, methods...) be called `allowCreateToken` or `tokenCreatePermitted` instead? The current name is a bit of a mouthful.

File: community/kernel/src/main/java/org/neo4j/procedure/Mode.java [hunk 6]
```diff
@@ -25,5 +25,6 @@
     WRITE,
     SCHEMA,
     DBMS,
-    DEFAULT
+    DEFAULT,
+    TOKEN
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/procedure/Mode.java]: This is the second enum called `Mode`, third if we include `AccessMode`. Could these be collapsed?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java [hunk 7]
```diff
@@ -242,6 +242,8 @@ private Mode mode( org.neo4j.procedure.Mode incoming )
             return Mode.DBMS;
         case SCHEMA:
             return Mode.SCHEMA_WRITE;
+        case TOKEN:
+            return Mode.TOKEN;
         case WRITE:
             return Mode.READ_WRITE;
         default:
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java]: This mapping smells.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java [hunk 8]
```diff
@@ -1524,16 +1541,40 @@ public void unregisterExecutingQuery( ExecutingQuery executingQuery )
...
+        return callProcedure( name, input, new RestrictedAccessMode( tx.securityContext().mode(), write ) );
     }
 
     @Override
     public RawIterator<Object[],ProcedureException> procedureCallWriteOverride( QualifiedName name, Object[] input )
             throws ProcedureException
     {
+        AccessMode write =
+                procedures.isAllowWriteTokenCreate() ? AccessMode.Static.TOKEN_WRITE : AccessMode.Static.WRITE;
+        return callProcedure( name, input, new OverriddenAccessMode( tx.securityContext().mode(), write ) );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java]: Could `isAllowWriteTokenCreate()` directly return the `AccessMode` we need here? As a "baseline access mode"?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/Procedures.java (hunk 5 above)
Comment (chrisvest) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/proc/Procedures.java]: Also, `Procedures` is exposed through `DepedencyResolver` in `DataSourceModule`, so user provided procedures could potentially change this setting.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseSecurityModule.java [hunk 9]
```diff
@@ -91,6 +91,9 @@ public void setup( Dependencies dependencies ) throws KernelException
             );
         life.add( securityLog );
 
+        Boolean allowTokenCreate = config.get( SecuritySettings.allow_publisher_create_token );
```
Comment (chrisvest) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseSecurityModule.java]: I'd prefer to see a lower-case `boolean` here, to make it clear that we really don't expect to see a `null` pop out of this setting. Also, it makes auto-unboxing only happen once.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedRolesBuilder.java [hunk 10]
```diff
@@ -28,14 +28,15 @@
...
 import static org.neo4j.server.security.enterprise.auth.plugin.api.PredefinedRoles.*;
 
 public class PredefinedRolesBuilder implements RolesBuilder
 {
     public static final Map<String,SimpleRole> roles = staticBuildRoles();
 
-    public static Map<String,SimpleRole> staticBuildRoles()
+    private static Map<String,SimpleRole> staticBuildRoles()
     {
         Map<String, SimpleRole> roles = new LinkedHashMap<>( 4 );
```
Comment (chrisvest) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedRolesBuilder.java]: Why `LinkedHashMap`? Why does this method put a _mutable_ map in a `public static final` field of a `public` class? I would recommend wrapping the returned map in `Collections.unmodifiableMap`, and have the inner map be a `ConcurrentHashMap` because `setAllowPublisherTokenCreate` can modify it at runtime.

Also, how does this being a static resource interact when multiple databases, with different configurations, are started up in the same Java process?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedRolesBuilder.java [hunk 11]
```diff
@@ -46,6 +47,7 @@
         SimpleRole architect = new SimpleRole( ARCHITECT );
         architect.add( new WildcardPermission( "schema:*" ) );
         architect.add( new WildcardPermission( "data:*" ) );
+        architect.add( new WildcardPermission( "token:*" ) );
```
Comment (chrisvest) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedRolesBuilder.java]: These tokens could be constants. Would slightly speed up the Set operations inside `SimpleRole` when things are added and deleted.

**Review by chrisvest — CHANGES_REQUESTED** [review]
I have done my first review round. It may be that I'm not terribly familiar with the security code, but I have a number of questions. And one or two things I think need more work.

**Review by chrisvest — APPROVED** [review]

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/Procedures.java [hunk 12]
```diff
@@ -217,4 +219,25 @@ public void start() throws Throwable
         // And register built-in procedures
         builtin.accept( this );
     }
+
+    private boolean changed = false;
+
+    public void writerCreateToken( boolean allow )
+    {
+        if ( !changed )
```
Comment (craigtaverner) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/Procedures.java:227]: This seems to be protecting against people abusing the dependency resolver to enable token writes from within procedures. However, that assumes access to GraphDatabaseAPI which already allows them to bypass security entirely, so we are perhaps protecting unnecessarily. Better to remove access to GraphDatabaseAPI. Although that can be part of another PR.

**Review by craigtaverner — APPROVED** [review]
